File: src/main/java/com/e16din/lightutils/utils/LogUtils.java
package com.e16din.lightutils.utils;

import android.util.Log;

import java.util.ArrayList;

/**
 * Created by e16din on 02.09.15.
 */
public class LogUtils extends SecureUtils {

    private static String tag = "LogUtils";

    private static ArrayList<String> blackList = new ArrayList<>();

    public static void disableLogs(String tag) {
        blackList.add(tag);
    }

    public static void disableLogs(Class cls) {
        disableLogs(cls.getSimpleName());
    }

    public static void enableLogs(String tag) {
        blackList.remove(tag);
    }

    public static void enableLogs(Class cls) {
        enableLogs(cls.getSimpleName());
    }

    public static void setTag(String tag) {
        LogUtils.tag = tag;
    }

    public static String getTag() {
        return tag;
    }


    public static void beep() {
        for (String tag : blackList) {
            if (tag.equals(LogUtils.tag))
                return;
        }
        Log.d(tag, "beep beep!");
    }

    public static void beep(Class cls) {
        for (String tag : blackList) {
            if (tag.equals(cls.getSimpleName()))
                return;
        }
        d(cls, "beep!");
    }

    public static void beep(String text) {
        for (String tag : blackList) {
            if (tag.equals(LogUtils.tag))
                return;
        }
        Log.d(tag, text);
    }

    public static void log(String text) {
        for (String tag : blackList) {
            if (tag.equals(LogUtils.tag))
                return;
        }
        Log.d(tag, text);
    }

    public static void d(Class cls, String text) {
        for (String tag : blackList) {
            if (tag.equals(cls.getSimpleName()))
                return;
        }
        Log.d(cls.getSimpleName(), text);
    }

    public static void d(String text) {
        for (String tag : blackList) {
            if (tag.equals(LogUtils.tag))
                return;
        }

        Log.d(tag, text);
    }

    public static void e(Class cls, String text) {
        for (String tag : blackList) {
            if (tag.equals(cls.getSimpleName()))
                return;
        }
        Log.e(cls.getSimpleName(), text);
    }

    public static void e(String text) {
        for (String tag : blackList) {
            if (tag.equals(LogUtils.tag))
                return;
        }

        Log.e(tag, text);
    }

    public static void i(Class cls, String text) {
        for (String tag : blackList) {
            if (tag.equals(cls.getSimpleName()))
                return;
        }

        Log.i(cls.getSimpleName(), text);
    }

    public static void i(String text) {
        for (String tag : blackList) {
            if (tag.equals(LogUtils.tag))
                return;
        }

        Log.i(tag, text);
    }

    public static void v(Class cls, String text) {
        for (String tag : blackList) {
            if (tag.equals(cls.getSimpleName()))
                return;
        }

        Log.v(cls.getSimpleName(), text);
    }

    public static void v(String text) {
        for (String tag : blackList) {
            if (tag.equals(LogUtils.tag))
                return;
        }

        Log.v(tag, text);
    }

    public static void w(Class cls, String text) {
        for (String tag : blackList) {
            if (tag.equals(cls.getSimpleName()))
                return;
        }

        Log.w(cls.getSimpleName(), text);
    }

    public static void w(String text) {
        for (String tag : blackList) {
            if (tag.equals(LogUtils.tag))
                return;
        }

        Log.w(tag, text);
    }

    public static void wtf(Class cls, String text) {
        for (String tag : blackList) {
            if (tag.equals(cls.getSimpleName()))
                return;
        }

        Log.wtf(cls.getSimpleName(), text);
    }

    public static void wtf(String text) {
        for (String tag : blackList) {
            if (tag.equals(LogUtils.tag))
                return;
        }

        Log.wtf(tag, text);
    }
}

File: src/main/java/com/e16din/lightutils/utils/LogUtils.kt
package com.e16din.lightutils.utils

import android.util.Log
import com.e16din.lightutils.LightUtils
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

enum class LogType { D, I, W, E }

const val TAG_DEBUG = "debug"

fun String?.logD(tag: String = "${TAG_DEBUG}D") {
    this.log(tag, LogType.D)
}

fun String?.logI(tag: String = "${TAG_DEBUG}I") {
    this.log(tag, LogType.I)
}

fun String?.logW(tag: String = "${TAG_DEBUG}W") {
    this.log(tag, LogType.W)
}

fun String?.logE(tag: String = "${TAG_DEBUG}E") {
    this.log(tag, LogType.E)
}

fun String?.log(tag: String = TAG_DEBUG, type: LogType = LogType.D) {
    LogUtils.log(this, tag, type)
}


open class LogUtils : TextViewUtils() {
    companion object {

        private const val LOG_PART_SIZE = 4000

        private const val DEFAULT_INDENT_SPACES = 4

        @JvmStatic
        fun log(obj: Any?, tag: String = TAG_DEBUG, type: LogType = LogType.D) {
            val isNull = obj != null
            logJson(isNull.then("$obj", "log object is null!"),
                    tag,
                    type = isNull.then(type, LogType.E))
        }

        @JvmStatic
        fun logJson(message: String?,
                    tag: String = TAG_DEBUG,
                    indentSpaces: Int = DEFAULT_INDENT_SPACES,
                    type: LogType = LogType.D) {

            if (!LightUtils.isDebug) return // else {

            if (!message.isNullOrEmpty() && message!!.startsWith("{")) {
                val jsonObject: JSONObject
                try {
                    jsonObject = JSONObject(message)
                    val logStr = jsonObject.toString(indentSpaces)

                    logLong(logStr, tag, type)
                } catch (e: JSONException) {
                    e.printStackTrace()

                    logLong("$message", tag, LogType.E)
                }

            } else if (!message.isNullOrEmpty() && message!!.startsWith("[")) {
                val jsonArray: JSONArray
                try {
                    jsonArray = JSONArray(message)

                    log("[", tag, type)
                    (0 until jsonArray.length())
                            .map { jsonArray.get(it) as JSONObject }
                            .forEach { logLong(it.toString(indentSpaces), tag, type) }
                    log("]", tag, type)

                } catch (e: JSONException) {
                    e.printStackTrace()

                    logLong("$message", tag, LogType.E)
                }

            } else {
                logLong(message, tag, type)
            }
        }

        @JvmStatic
        fun logLong(message: String?, tag: String = TAG_DEBUG, type: LogType = LogType.D) {
            if (!LightUtils.isDebug) return

            if (message != null && message.length > LOG_PART_SIZE) {
                val nextPart = message.substring(0, LOG_PART_SIZE)
                log(nextPart, tag, type)

                logLong(tag, message.substring(LOG_PART_SIZE), type)

            } else {
                log(message, tag, type)
            }
        }

        @JvmStatic
        private fun log(message: String?, tag: String, type: LogType) {
            when (type) {
                LogType.D -> Log.d(tag, message)
                LogType.I -> Log.i(tag, message)
                LogType.W -> Log.w(tag, message)
                LogType.E -> Log.e(tag, message)
            }
        }
    }
}
